src/main/java/appeng/items/tools/MemoryCardItem.java

```diff
@@ -59,6 +75,200 @@ public MemoryCardItem(Item.Properties properties) {
+                desiredUpgrades.put(item, desiredCount);
+            }
+        }
+
+        var upgrades = upgradeableObject.getUpgrades();
+
+        // In creative mode, just set it exactly as the memory card says
+        if (player.getAbilities().instabuild) {
+            // Clear it out first
+            for (int i = 0; i < upgrades.size(); i++) {
+                upgrades.setItemDirect(i, ItemStack.EMPTY);
+            }
+            for (var entry : desiredUpgrades.entrySet()) {
+                upgrades.addItems(new ItemStack(entry.getKey(), entry.getValue()));
+            }
+            return true;
+        }
+
+        var upgradeSources = new ArrayList<InternalInventory>();
+        upgradeSources.add(new PlayerInternalInventory(player.getInventory()));
+
+        // Search the player for a network tool
+        var networkTool = NetworkToolItem.findNetworkToolInv(player);
+        if (networkTool != null) {
+            upgradeSources.add(networkTool.getInventory());
+        }
+
+        // Move out excess
+        for (int i = 0; i < upgrades.size(); i++) {
+            var current = upgrades.getStackInSlot(i);
+            if (current.isEmpty()) {
+                continue;
+            }
+
+            var desiredCount = desiredUpgrades.getOrDefault(current.getItem(), 0);
+            var totalInstalled = upgradeableObject.getInstalledUpgrades(current.getItem());
+            var toRemove = totalInstalled - desiredCount;
+            if (toRemove > 0) {
+                var removed = upgrades.extractItem(i, toRemove, false);
+
+                for (var upgradeSource : upgradeSources) {
+                    if (!removed.isEmpty()) {
+                        removed = upgradeSource.addItems(removed);
+                    }
+                }
+                if (!removed.isEmpty()) {
+                    player.drop(removed, false);
+                }
+            }
+        }
+
+        // Move in what's still missing
+        for (var entry : desiredUpgrades.entrySet()) {
+            var missingAmount = entry.getValue() - upgradeableObject.getInstalledUpgrades(entry.getKey());
+            if (missingAmount > 0) {
+                var potential = new ItemStack(entry.getKey(), missingAmount);
+                // Determine how many can we *actually* insert
+                var overflow = upgrades.addItems(potential, true);
+                if (!overflow.isEmpty()) {
+                    missingAmount -= overflow.getCount();
+                }
+
+                // Try getting them from the network tool or player inventory
+                for (var upgradeSource : upgradeSources) {
+                    var cards = upgradeSource.removeItems(missingAmount, potential, null);
+                    if (!cards.isEmpty()) {
+                        overflow = upgrades.addItems(cards);
+                        if (!overflow.isEmpty()) {
+                            player.getInventory().placeItemBackInInventory(overflow);
```

Comment:
Not a fan of this as it silently moves the items around in the inventory afaict.